workflow/docs/design.md

```diff
@@ -0,0 +1,406 @@
+}
+```
+
+- `Notify` as `Enqueue`
+- `Acquire` as `Dequeue`
+
+`OperableTrigger` is the bridge between workflow engine and other components.
+
+> For implementation with kubernetes, we should make a **Trigger** which is a reconciler in the controller-manager.
+
+### Scheduler
+
+Scheduler should pick out one or more templates that should be executed next. Scheduler should:
+
+- parse the whole Workflow(or other components tell scheduler the hierarchy of workflow, I'm not sure for that)
+- pick next template(s) to run
+- evaluate expressions for conditional branch(only for `Task` Template, which contains conditional branches)
+
+We should implement at least 4 types of Scheduler:
+
+- `EntryScheduler`, it's the first scheduler for each workflow, it only picks the `entry` template.
+- `SerialScheduler`, it should pick the next **one** template from children templates in SerialTemplate.
+- `ParallelScheduler`, it should pick the all templates from children templates in ParallelTemplate.
+- `TaskScheduler`, it is based on `ParallelScheduler`, but contains logic about evaluation expressions for conditional branches.
+
+```go
+type Scheduler interface{
+  ScheduleNext(ctx context.Context) (nextTemplates []template.Template, parentNodeName string, err error)
+}
+```
+
+### Actor and Playground
+
+> - Nothing happens until the Actor works.
+> - The status of Workflow does NOT present the result for the Actor. (It means we could not fetch the result of one "Actor" from kubernetes.) Because the "real world" is so complicated to describe, we want the result of "Actor" do not "pollution" our schedule objects.
+> - While an "Actor" failed, the whole workflow instance is trapped into failed.
+
+The Actor will do all "dirty" effects, for example: creating Chaos Resource, running user-defined tasks, and so on.
+
+```go
+type Actor interface {
+  PlayOn(pg Playground) error
+}
+
+type Playground interface {
+  CreateNetworkChaos(networkChaos chaosmeshv1alph1.NetworkChaos) error
+  DeleteNetworkChaos(namespace, name string) error
+  CreatePodChaos(podChaos chaosmeshv1alph1.PodChaos) error
+  DeletePodChaos(namespace, name string) error
+
+  // Other ad-hoc methods
+}
+```
+
+### WorkflowManager: rest glue codes
+
+We need a manager who makes these components work together, and controls the main-loop, and update the status of the workflow, and max-concurrency, and so-on.
+
+The main loop looks like:
+
+```go
+func loop(){
+  for {
+    workflowKey, err := trigger.AcquireNext()
+    scheduler, err := fetchSchedulerFor(workflowKey)
+
+    go func(){
+      templates, err := scheduler.scheduleNext()
+      for item := range templates {
+        go func (){
+          // TODO: update node status
+          actor := fetchActor(item)
+          err := actor.Apply()
+          // TODO: update node status
+          if item.NeedReQueue() {
+            // TODO: manually ask for next trigger
+          }
+        }()
+      }
+    }
+  }
+}
+```
+
+### States for Node
```

Comment:
This design combines logic tasks, chaos tasks, and etc into one state machine, which makes the state machine more complicated. How about splitting them into two or more state machines?  Such as logic node machine and execute node machine.

Reply:
Yes, it is just like a mess. We should reconsider how to recognize codes about different types of state-machine.

I will append this part in the next PRs.